oracle/src/events/processHalfDuplexTransfers/index.js

```diff
@@ -0,0 +1,132 @@
+require('../../../env')
+const promiseLimit = require('promise-limit')
+const { HttpListProviderError } = require('http-list-provider')
+const { BRIDGE_VALIDATORS_ABI, ZERO_ADDRESS } = require('../../../../commons')
+const rootLogger = require('../../services/logger')
+const { web3Home, web3Foreign } = require('../../services/web3')
+const { AlreadyProcessedError, AlreadySignedError, InvalidValidatorError } = require('../../utils/errors')
+const { EXIT_CODES, MAX_CONCURRENT_EVENTS } = require('../../utils/constants')
+const estimateGas = require('../processAffirmationRequests/estimateGas')
+
+const limit = promiseLimit(MAX_CONCURRENT_EVENTS)
+
+let validatorContract = null
+
+function processTransfersBuilder(config) {
+  const homeBridge = new web3Home.eth.Contract(config.homeBridgeAbi, config.homeBridgeAddress)
+  const foreignBridge = new web3Foreign.eth.Contract(config.foreignBridgeAbi, config.foreignBridgeAddress)
+  const userRequestForAffirmationAbi = config.foreignBridgeAbi.filter(
+    e => e.type === 'event' && e.name === 'UserRequestForAffirmation'
+  )[0]
+  const tokensSwappedAbi = config.foreignBridgeAbi.filter(e => e.type === 'event' && e.name === 'TokensSwapped')[0]
+  const userRequestForAffirmationHash = web3Home.eth.abi.encodeEventSignature(userRequestForAffirmationAbi)
+  const tokensSwappedHash = tokensSwappedAbi ? web3Home.eth.abi.encodeEventSignature(tokensSwappedAbi) : '0x'
+
+  return async function processTransfers(transfers, blockNumber) {
+    const txToSend = []
+
+    if (validatorContract === null) {
+      rootLogger.debug('Getting validator contract address')
+      const validatorContractAddress = await homeBridge.methods.validatorContract().call()
+      rootLogger.debug({ validatorContractAddress }, 'Validator contract address obtained')
+
+      validatorContract = new web3Home.eth.Contract(BRIDGE_VALIDATORS_ABI, validatorContractAddress)
+    }
+
+    rootLogger.debug(`Processing ${transfers.length} Transfer events`)
+    const callbacks = transfers
+      .map(transfer => async () => {
+        const { from, value } = transfer.returnValues
+
+        const logger = rootLogger.child({
+          eventTransactionHash: transfer.transactionHash
+        })
+
+        logger.info({ from, value }, `Processing transfer ${transfer.transactionHash}`)
+
+        const block = await web3Foreign.eth.getBlock(blockNumber)
```

Comment:
Do you think it makes sense to report the block timestamp in the log?

Reply:
Yes, added timestamp log https://github.com/poanetwork/tokenbridge/pull/244/commits/ab2c0ea120fa1efe6af213ce9400b7cb0a291bee